File: lista9/darkroads.py
#disjoint set union
def init(m):
    for i in range(m):
        parent[i] = i
        size[i] = 1

def find(a):
    if parent[a] == a: return a
    parent[a] = find(parent[a])
    return parent[a]

def union(a,b):
    a,b = find(a),find(b)
    if a != b:
        if size[a] < size[b]: a,b = b,a
        parent[b] = a
    size[a] += size[b]

maxLimit = int(2e5)
parent = [0] * maxLimit # ancestral do vertice
size = [0] * maxLimit # tamanho do componente q o vertice esta inserido
m,n = map(int,input().split())

#kruskal
while m != 0:
    init(m)
    edges = []
    mstSum = 0
    graphSum = 0
    for i in range(n):
        u,v,w = map(int,input().split())
        edges.append((w,u,v))
    edges.sort()
    for w,u,v in edges:
        u,v = find(u),find(v)
        graphSum += w
        if u != v:
            mstSum += w
            union(u,v)
    maxEconomy = graphSum - mstSum
    print(maxEconomy)
    m,n = map(int,input().split())

File: lista9/darkroads.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

const int maxLimit = 2e5;
vector<int> parent(maxLimit); // Ancestral do vertice
vector<int> componentSize(maxLimit);   // Tamanho do componente que o vertice esta inserido

void init(int m) {
    for (int i = 0; i < m; i++) {
        parent[i] = i;
        componentSize[i] = 1;
    }
}

int find(int a) {
    if (parent[a] == a)
        return a;
    parent[a] = find(parent[a]);
    return parent[a];
}

void unionSets(int a, int b) {
    a = find(a);
    b = find(b);
    if (a != b) {
        if (componentSize[a] < componentSize[b])
            swap(a, b);
        parent[b] = a;
        componentSize[a] += componentSize[b];
    }
}

int main() {
    int m, n;
    while (cin >> m && m != 0) {
        init(m);
        vector<pair<int, pair<int, int>>> edges;
        int mstSum = 0;
        int graphSum = 0;

        cin >> n;

        for (int i = 0; i < n; i++) {
            int u, v, w;
            cin >> u >> v >> w;
            edges.push_back({w, {u, v}});
        }

        sort(edges.begin(), edges.end());

        for (const auto& edge : edges) {
            int w = edge.first;
            int u = edge.second.first;
            int v = edge.second.second;
            u = find(u);
            v = find(v);
            graphSum += w;

            if (u != v) {
                mstSum += w;
                unionSets(u, v);
            }
        }

        int maxEconomy = graphSum - mstSum;
        cout << maxEconomy << endl;
    }

    return 0;
}